app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-form-change-tracker.js

```diff
@@ -0,0 +1,153 @@
+window.GOVUK = window.GOVUK || {}
+window.GOVUK.Modules = window.GOVUK.Modules || {};
+
+(function (Modules) {
+  'use strict'
+
+  function Ga4FormChangeTracker (module) {
+    this.module = module
+  }
+
+  Ga4FormChangeTracker.prototype.init = function () {
+    var consentCookie = window.GOVUK.getConsentCookie()
+
+    if (consentCookie && consentCookie.usage) {
+      this.startModule()
+    } else {
+      this.start = this.startModule.bind(this)
+      window.addEventListener('cookie-consent', this.start)
+    }
+  }
+
+  // triggered by cookie-consent event, which happens when users consent to cookies
+  Ga4FormChangeTracker.prototype.startModule = function () {
+    if (window.dataLayer) {
+      window.removeEventListener('cookie-consent', this.start)
+      this.module.addEventListener('change', this.trackFormChange.bind(this))
+    }
+  }
+
+  Ga4FormChangeTracker.prototype.getJson = function (target, attribute) {
+    let dataContainer
+    let data
+
+    try {
+      dataContainer = target.closest(`[${attribute}]`)
+      data = dataContainer.getAttribute(attribute)
+      return JSON.parse(data)
+    } catch (e) {
+      console.error(
+        `GA4 configuration error: ${e.message}, attempt to access ${attribute} on ${target}`,
+        window.location
+      )
+    }
+  }
+
+  Ga4FormChangeTracker.prototype.dateTimeComponent = function (target) {
+    return (
+      target.closest('.app-c-datetime-fields') ||
+      target.closest('.govuk-date-input')
+    )
+  }
+
+  Ga4FormChangeTracker.prototype.getSection = function (target, checkableValue) {
```

Comment:
This function seems quite complicated in terms of things looking for things around the DOM and the necessary error checking surrounding that. Is there perhaps a cleaner approach where we could set attributes on specific elements and target those instead?

I'm guessing the reasoning behind this is that this tracker can be applied to existing forms without having to modify them?

Reply:
I've re-written it somewhat now. I've removed the logic for handling the `datetime_fields` component and replaced it with checking for a `ga4-form-section` attribute.